src/dbnode/x/xio/segment_reader.go

```diff
@@ -46,40 +47,101 @@ func (sr *segmentReader) Clone(
 	return NewSegmentReader(sr.segment.Clone(pool)), nil
 }
 
-func (sr *segmentReader) Read(b []byte) (int, error) {
-	if len(b) == 0 {
-		return 0, nil
+func (sr *segmentReader) Read64() (word uint64, n byte, err error) {
+	sr.lazyInit()
+
+	var (
+		nh    = len(sr.lazyHead)
+		nht   = nh + len(sr.lazyTail)
+		res   uint64
+		bytes byte
+	)
+
+	if sr.si >= nht {
+		return 0, 0, io.EOF
 	}
 
-	if b := sr.segment.Head; b != nil && len(sr.lazyHead) == 0 {
-		sr.lazyHead = b.Bytes()
+	if sr.si+8 < nh {
+		// NB: this compiles to a single 64 bit load followed by
+		// a BSWAPQ on amd64 gc 1.13 (https://godbolt.org/z/oTK1jx).
+		res = binary.BigEndian.Uint64(sr.lazyHead[sr.si:])
+		sr.si += 8
+		return res, 8, nil
 	}
-	if b := sr.segment.Tail; b != nil && len(sr.lazyTail) == 0 {
-		sr.lazyTail = b.Bytes()
+
+	if sr.si < nh {
+		for ; sr.si < nh; sr.si++ {
+			res = (res << 8) | uint64(sr.lazyHead[sr.si])
+			bytes++
+		}
+		for ; sr.si < nht && bytes < 8; sr.si++ {
+			res = (res << 8) | uint64(sr.lazyTail[sr.si-nh])
+			bytes++
+		}
+		return res << (64 - 8*bytes), bytes, nil
 	}
 
-	nh, nt := len(sr.lazyHead), len(sr.lazyTail)
-	if sr.si >= nh+nt {
-		return 0, io.EOF
+	if sr.si+8 < nht {
+		// NB: this compiles to a single 64 bit load followed by
+		// a BSWAPQ on amd64 gc 1.13 (https://godbolt.org/z/oTK1jx).
+		res = binary.BigEndian.Uint64(sr.lazyTail[sr.si-nh:])
+		sr.si += 8
+		return res, 8, nil
 	}
-	n := 0
-	if sr.si < nh {
-		nRead := copy(b, sr.lazyHead[sr.si:])
-		sr.si += nRead
-		n += nRead
-		if n == len(b) {
-			return n, nil
+
+	for ; sr.si < nht && bytes < 8; sr.si++ {
```

Comment:
I think `bytes < 8` can be dropped (just like in [the `for` on `lazyHead`](https://github.com/m3db/m3/pull/2827/files#diff-3e6c8bee4811cf296ffb0bb8567b3272a641d696b0cc348e4a2a2984e6fa6c08R73-R76)), since otherwise previous `if` should have matched.

```suggestion
	for ; sr.si < nht; sr.si++ {
```

Reply:
Makes sense, fixing.